### PR TITLE
Issue-123 - Change uncommon-spacing handling from leading to trailing…

### PR DESCRIPTION
**Pull request by DoubleDee73**
- Adjusted Auto-Correct for uncommon spacing from leading spaces to trailing spaces
- Adjusted "Rolling Left/Right" feature to accomodate spacing changes
- Adjusted "Splitting Note" feature

### CODE REVIEW

**Comment by Daniel20Ultrastar-es** [comment]
Thousands and thousands of songs are with the space in front of them. I don't find it an advantage to change it. 

If you put the space afterwards you can't tell at a glance if you open the document if there is a space or not, because the space is a hidden character.

In programming the characters on the right are considered errors and some wizards delete those characters as cleanup, so it is not a good idea to put whitespace on the right.

**Comment by DoubleDee73** [comment]
Hi @Daniel20Ultrastar-es !
This has been a discussion around the community for quite a while now. As far as I can tell, it has concluded in favour of trailing spaces, as the OG piece of software (namely Ultrastar Deluxe) has changed its Song Editor to auto correct to this way of spacing. Other tools like "Ultrastar Manager" or "Ultrastar Creator" which also ship with USDX have now also changed to trailing spaces.
Personally, I don't prefer any method of the other (and I was not involved in that discussion either, so I don't necessarily want to reopen it either). But since the main active applications have settled for this method, I don't see a reason, why Yass should not adapt. In any event, there is no harm you can't leave your existing songs the way they are (I'd say half my library is still the "old way"). 

Having said that, when I create new songs, and I want used Yass to finetune, I found it extremely annoying, that all my notes used to be red, because of "uncommon spacing" and I couldn't see my golden notes anymore.

In the end it's @SarutaSan72 's decision to merge or not. If he decides not to, I'll just keep my fork.

**Comment by bohning** [comment]
I fully support @DoubleDee73’s reasoning and support this change.

**Comment by TeLiXj** [comment]
I think that this change is a mistake. I can't see **any** advantage to using trailing spaces, only disadvantages as Daniel said. 
I they changed on their own that's their problem, they can use their own editor.

**Comment by SarutaSan72** [comment]
Whitespaces are placed in front since Corvus' first version, which was afaik before 2009. I don't think that this should be changed, too many existing songs are perfectly finetuned with this. Too many longtime users would be irritated.

The current pull request would change Yass' default behavior. I can't do that, all existing songs that were errorfree in Yass would then be marked with errors. And possibly batch-corrected, whew. You could add the alternative behavior as option. But maybe this would create more confusion and incompatibilies. Apart from bugs and required testing. 

I have very limited time and energy, and personally I feel this isn't worth it.

**Comment by TeLiXj** [comment]
> I would have never guessed, that this is such a touchy topic. My guess is, that the majority of users won't even notice.
> 
> As for advantages:
> 
>     1. The major tooling applications, that are actually maintained have changed the way syllables are being spaced
> 
>     2. When you are manually editing TXTs and you are removing or adding a page break, you can do so without having to add or remove spaces
> 
>     3. It's not consistent, when words start with a space, but at the beginning of a page they do not. From a developing point of view, you always have to deal with it separately ("if (syllable at PageStart) { do this } else { do that }
> 
>     4. That's just nitpicking, but in the USDX editor, the first word never has a margin, whereas all following words have the noticable space at the beginning
> 
> 
> I'm thinking, that moving forward, all the software that revolves around Ultrastar should deal with these things in a similar way.
> 
> By the way, I just stumbled across this other cool project that automatically creates Ultrastar-TXT files https://github.com/rakuri255/UltraSinger It's still in development, but the idea is, you put in a source, and it will detect the language, the lyrics, split the syllables, split the audio into instrumental + vocals, detect the pitch and create a txt. And that one also adds the spaces at the end...

1. Only a part of UltraStar, this isn't a reason to change it. Yass is **THE** tool to edit songs since years.
2. Irrelevant, it's not the main use of the manual editing.
3. If this is a problem (which I think it isn't) I prefer to add a space to the first word. Maybe it's easier to manage the initial page break.
4. Return to point 1

We will follow your new project!! Sounds great!!!

**Comment by DoubleDee73** [comment]
@TeLiXj Hang on a sec. 
Concerning 1. Says who? I have created numerous songs and edited even more. While I do use Yass for fine-tuning, it definitely is not my starting point. I'm using Ultrastar Creator for tapping and the USDX Creator for pitching. (Especially due to https://github.com/SarutaSan72/Yass/issues/99)
In our community, 47% are using Yass and 42% USDX Editor. While there is the majority in terms of editing, there is the question of workflow as you probably won't use Yass for tapping. And it's not like the resounding majority either...

You guys are talking about disadvantages. I only counted two. 
1. "If you put the space afterwards you can't tell at a glance if you open the document if there is a space or not, because the space is a hidden character"
But you also said "Irrelevant, it's not the main use of the manual editing". 
2. "In programming the characters on the right are considered errors/comments and some wizards delete those characters as cleanup, so it is not a good idea to put whitespace on the right."
I don't see how this is an issue. If you don't proactive trim each line programmatically, nothing will be changed. That's why there are tools.

As far as I understood, @SarutaSan72 doesn't want to change it now anyway (I don't understand why this was an open issue then), you do

**Comment by SarutaSan72** [comment]
> I don't understand why this was an open issue then
because it needed discussion and actual code analysis. so the above is "progress" (although way too emotional, which is why I was not reading it earlier) --> closing it with the option to reopen and include it later "as an option"
